### PR TITLE
deprecate spack diy in favor of spack dev-build

### LINKED ISSUE #3642: spack configure/build doesn't work for all package base classes
`spack configure` and `spack build` are designed stop at a certain phase of the package installation process for debugging purposes. However, this only works for `CMakePackage`, `AutotoolsPackage`, and `PythonPackage`. These 3 build systems contain separate build and install phases.

This doesn't work for `MakefilePackage` even though it has a build phase:
```
$ spack build lammps
==> Error: Package lammps has no build phase, or its build phase is not separated from install
```
It also doesn't work for a `Package` class with manually added build phases:
```
$ spack build qbank
==> Error: Package qbank has no build phase, or its build phase is not separated from install
```
See #3641 for this qbank package.

Based on the commands, it looks like we are looking for the package base class in `build_systems_to_phase`, a dictionary that converts the phase to the name of the function. If the base class isn't in this dictionary, it crashes, even if the package has a build phase defined.

@alalazo

### PR DESCRIPTION
**Pull request by becker33**
Fixes #3642.

We've decided to rename the `spack diy` command to `spack dev-build` to make the use case clearer.

The `spack diy` command has some useful functionality for developers using Spack to build their dependencies and configure/build/install the code they are developing. We've realized that as this becomes a developer tool, the name `spack diy` is less useful. Hence the rename.

The `spack dev-build` command has a `-u/--until PHASE` option to stop after a given phase of the build. This can be used to configure your project, run cmake on your project, or similarly stop after any stage of the build the user wants. These options are analogous to the existing `spack configure` and `spack build` commands, but for developer builds. To unify the syntax, we have deprecated the `spack configure` and `spack build` commands, and added a `-u/--until PHASE` option to the `spack install` command.

I believe the functionality in `spack dev-build` (specifically `spack dev-build -u cmake`) can supercede the `spack setup` command, but this PR does not deprecate that command as that will require slightly more thought.

### CODE REVIEW

**Comment by adamjstewart** [comment]
Will this fix #3642? That is, can I use `spack install --until foo` where `foo` is an arbitrary phase defined in a custom `Package` class?

**Comment by tgamblin** [comment]
@adamjstewart: yes

File: lib/spack/spack/cmd/build.py [hunk 1]
```diff
@@ -38,4 +39,6 @@ def setup_parser(subparser):
 
 
 def build(parser, args):
+    tty.warn("This command is deprecated. Use `spack install --until` to"
```
Comment (tgamblin) [inline lib/spack/spack/cmd/build.py]: Add that this will be removed in a future Spack version.

File: lib/spack/spack/cmd/configure.py [hunk 2]
```diff
@@ -82,4 +82,6 @@ def _stop_at_phase_during_install(args, calling_fn, phase_mapping):
 
 
 def configure(parser, args):
+    tty.warn("This command is deprecated. Use `spack install --until` to"
```
Comment (tgamblin) [inline lib/spack/spack/cmd/configure.py]: Add that this will be removed in a future Spack version.

File: lib/spack/spack/cmd/diy.py [hunk 3]
```diff
@@ -50,6 +50,7 @@ def setup_parser(subparser):
 
 
 def diy(self, args):
+    tty.warn("`spack diy` is deprecated in favor of `spack dev-build`.")
```
Comment (tgamblin) [inline lib/spack/spack/cmd/diy.py]: This should really say that all the functionality of *diy* has *moved* to `dev-build`, and that the `diy` command will be removed in a future Spack version.

To remove redundant code, I would recommend that we implement `diy` the same way `spack compilers` is implemented -- that is, delete all the code here and have the parser setup and the code in this method call `dev-build`.  You can just import `spack.cmd.diy` and call its methods.

**Pull request merged**